src/storage/src/compaction/twcs.rs

```diff
@@ -0,0 +1,391 @@
+use common_time::Timestamp;
+use store_api::logstore::LogStore;
+
+use crate::compaction::picker::get_expired_ssts;
+use crate::compaction::task::CompactionOutput;
+use crate::compaction::{infer_time_bucket, CompactionRequestImpl, CompactionTaskImpl, Picker};
+use crate::sst::{FileHandle, LevelMeta};
+
+pub struct TwcsPicker<S> {
+    max_files_in_active_window: usize,
+    max_files_in_non_active_window: usize,
+    _phantom_data: PhantomData<S>,
+}
+
+impl<S> Debug for TwcsPicker<S> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TwcsPicker")
+            .field(
+                "max_files_in_active_window",
+                &self.max_files_in_active_window,
+            )
+            .field(
+                "max_files_in_non_active_window",
+                &self.max_files_in_non_active_window,
+            )
+            .finish()
+    }
+}
+
+impl<S> Default for TwcsPicker<S> {
+    fn default() -> Self {
+        Self {
+            max_files_in_active_window: 4,
+            max_files_in_non_active_window: 1,
+            _phantom_data: Default::default(),
+        }
+    }
+}
+
+impl<S> TwcsPicker<S> {
+    /// Builds compaction output from files.
+    /// For active writing window, we allow for at most `max_files_in_active_window` files to alleviate
+    /// fragmentation. For other windows, we allow at most 1 file at each window.
+    fn build_output(
+        &self,
+        time_windows: &BTreeMap<i64, Vec<FileHandle>>,
+        active_window: Option<i64>,
+        window_size: i64,
+    ) -> Vec<CompactionOutput> {
+        let mut output = vec![];
+        for (window, files) in time_windows {
+            if let Some(active_window) = active_window && *window == active_window {
+                if files.len() > self.max_files_in_active_window {
+                    output.push(CompactionOutput {
+                        output_level: 1, // we only have two levels and always compact to l1 
+                        time_window_bound: *window,
+                        time_window_sec: window_size,
+                        inputs: files.clone(),
+                        // Strict window is not needed since we always compact many files to one 
+                        // single file in TWCS.
+                        strict_window: false,
+                    });
+                } else {
+                    debug!("Active window not present or no enough files in active window {:?}", active_window);
+                }
+            } else {
+                // not active writing window
+                if files.len() > self.max_files_in_non_active_window {
+                    output.push(CompactionOutput {
+                        output_level: 1,
+                        time_window_bound: *window,
+                        time_window_sec: window_size,
+                        inputs: files.clone(),
+                        strict_window: false,
```

Comment:
issue(non-blocking): Currently, TWCS might produce overlapping L1 files.

Suppose L0 files `a`, `b`, and `c` with the following ranges:
```
a: [4, 6]
b: [6, 8]
c: [3, 4]
```
The time window is 5, which is large enough. Then during compaction, we merge `a` and `b` according to their max timestamps and get `d`
```
a + b -> d: [4, 8]
```

Then we still need to merge `d` and `c` while scanning.

Reply:
That is a by-design caveat, just like cassandra. 

In "real-time workload", usually we don't expected a large amount of out-of-order insertions or updates. Thus TWCS won't generates heavily overlapping SST files. That's also why we use the max timestamp to determine the time window -- min timestamp would be some out-of-order requests.

Cassandra uses a *major compaction*, which is in fact a size-tiered compaction to ensure only one SST file per time window. 

But we don't have to introduce a major compaction, as long as we can split memtables strictly by time window during flush, then we won't end up with overlapping SST files during TWCS compaction.